README.md
[![pub package](https://img.shields.io/pub/v/srp.svg?color=important)](https://pub.dev/packages/srp)
---

# srp

Secure Remote Password for dart &amp; flutter

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

JavaScript version: https://github.com/LinusU/secure-remote-password


CHANGELOG.md
## 1.0.0

* first publish

## 1.0.1

* crypto: ^2.1.4 -> ^2.1.3

test/srp_test.dart
import 'package:srp/client.dart' as client;
import 'package:srp/server.dart' as server;
import 'package:test/test.dart';

void main() {
  test('srp test', () {
    final username = 'x@vipycm.com';
    final password = 'password';

    final salt = client.generateSalt();
    final privateKey = client.derivePrivateKey(salt, username, password);
    final verifier = client.deriveVerifier(privateKey);

    final clientEphemeral = client.generateEphemeral();
    final serverEphemeral = server.generateEphemeral(verifier);

    final clientSession = client.deriveSession(clientEphemeral.secret, serverEphemeral.public, salt, username, privateKey);
    final serverSession = server.deriveSession(serverEphemeral.secret, clientEphemeral.public, salt, username, verifier, clientSession.proof);

    client.verifySession(clientEphemeral.public, clientSession, serverSession.proof);

    expect(clientSession.key, serverSession.key);
  });
}


lib/params.dart
import 'dart:convert';
import 'package:convert/convert.dart';
import 'package:crypto/crypto.dart';
import './bigint_ex.dart';

final _input = {
  'largeSafePrime': 'AC6BDB41 324A9A9B F166DE5E 1389582F AF72B665 1987EE07 FC319294' +
      '3DB56050 A37329CB B4A099ED 8193E075 7767A13D D52312AB 4B03310D' +
      'CD7F48A9 DA04FD50 E8083969 EDB767B0 CF609517 9A163AB3 661A05FB' +
      'D5FAAAE8 2918A996 2F0B93B8 55F97993 EC975EEA A80D740A DBF4FF74' +
      '7359D041 D5C33EA7 1D281E44 6B14773B CA97B43A 23FB8016 76BD207A' +
      '436C6481 F1D2B907 8717461A 5B9D32E6 88F87748 544523B5 24B0D57D' +
      '5EA77A27 75D2ECFA 032CFBDB F52FB378 61602790 04E57AE6 AF874E73' +
      '03CE5329 9CCC041C 7BC308D8 2A5698F3 A8D0C382 71AE35F8 E9DBFBB6' +
      '94B5C803 D89F7AE4 35DE236D 525F5475 9B65E372 FCD68EF2 0FA7111F' +
      '9E4AFF73',
  'generatorModulo': '02',
  'hashFunction': 'sha256',
  'hashOutputBytes': 256 ~/ 8
};

class Params {
  static final N = BigInt.parse(_input['largeSafePrime'].toString().replaceAll(' ', ''), radix: 16);
  static final g = BigInt.parse(_input['generatorModulo'].toString().replaceAll(' ', ''), radix: 16);
  static final k = H([N, g]);
  static final int hashOutputBytes = _input["hashOutputBytes"];

  static BigInt H(List args) {
    final output = new AccumulatorSink<Digest>();
    final input = sha256.startChunkedConversion(output);
    for (final arg in args) {
      if (arg is BigInt) {
        var hexStr = arg.toHex();
        if (!hexStr.length.isEven) {
          hexStr = '0$hexStr';
        }
        input.add(hex.decode(hexStr));
      } else {
        input.add(utf8.encode(arg));
      }
    }
    input.close();
    final hexStr = output.events.single.toString();
    output.close();

    return BigInt.parse(hexStr, radix: 16);
  }
}


lib/server.dart
import 'dart:math';
import 'package:convert/convert.dart';
import './bigint_ex.dart';
import './types.dart';
import './params.dart';

Ephemeral generateEphemeral(String verifier) {
  final N = Params.N;
  final g = Params.g;
  final k = Params.k;

  final v = BigInt.parse(verifier, radix: 16);

  final random = Random.secure();
  final randomHex = hex.encode(List<int>.generate(Params.hashOutputBytes, (i) => random.nextInt(256)));

  final b = BigInt.parse(randomHex, radix: 16);
  final B = (k * v + g.modPow(b, N)) % N;

  return Ephemeral(secret: randomHex, public: B.toHex());
}

Session deriveSession(String serverSecretEphemeral, String clientPublicEphemeral, String salt, String username, String verifier, String clientSessionProof) {
  final N = Params.N;
  final g = Params.g;
  final k = Params.k;
  final H = Params.H;

  final b = BigInt.parse(serverSecretEphemeral, radix: 16);
  final A = BigInt.parse(clientPublicEphemeral, radix: 16);
  final s = BigInt.parse(salt, radix: 16);
  final I = username;
  final v = BigInt.parse(verifier, radix: 16);

  final B = (k * v + g.modPow(b, N)) % N;

  if (A % N == 0) {
    throw Exception('the client sent an invalid public ephemeral');
  }

  final u = H([A, B]);

  final S = (A * (v.modPow(u, N))).modPow(b, N);

  final K = H([S]);

  final M = H([
    H([N]) ^ (H([g])),
    H([I]),
    s,
    A,
    B,
    K
  ]);

  final expected = M;
  final actual = BigInt.parse(clientSessionProof, radix: 16);

  if (actual != expected) {
    throw Exception('Client provided session proof is invalid');
  }

  final P = H([A, M, K]);

  return Session(key: K.toHex(), proof: P.toHex());
}


lib/client.dart
import 'dart:math';
import 'package:convert/convert.dart';
import './bigint_ex.dart';
import './types.dart';
import './params.dart';

generateSalt() {
  final random = Random.secure();
  return hex.encode(List<int>.generate(Params.hashOutputBytes, (i) => random.nextInt(256)));
}

derivePrivateKey(String salt, String username, String password) {
  final H = Params.H;
  final s = BigInt.parse(salt, radix: 16);
  final x = H([
    s,
    H(['$username:$password'])
  ]);

  return x.toHex();
}

deriveVerifier(String privateKey) {
  final N = Params.N;
  final g = Params.g;
  return g.modPow(BigInt.parse(privateKey, radix: 16), N).toHex();
}

Ephemeral generateEphemeral() {
  final N = Params.N;
  final g = Params.g;
  final random = Random.secure();
  final randomHex = hex.encode(List<int>.generate(Params.hashOutputBytes, (i) => random.nextInt(256)));

  final a = BigInt.parse(randomHex, radix: 16);
  final A = g.modPow(a, N);

  return Ephemeral(secret: randomHex, public: A.toHex());
}

Session deriveSession(String clientSecretEphemeral, String serverPublicEphemeral, String salt, String username, String privateKey) {
  final N = Params.N;
  final g = Params.g;
  final k = Params.k;
  final H = Params.H;

  final a = BigInt.parse(clientSecretEphemeral, radix: 16);
  final B = BigInt.parse(serverPublicEphemeral, radix: 16);
  final s = BigInt.parse(salt, radix: 16);
  final I = username;
  final x = BigInt.parse(privateKey, radix: 16);

  final A = g.modPow(a, N);

  if (B % N == 0) {
    throw Exception('The server sent an invalid public ephemeral');
  }

  final u = H([A, B]);

  final S = (B - (k * (g.modPow(x, N)))).modPow(a + (u * x), N);

  final K = H([S]);

  final M = H([
    H([N]) ^ (H([g])),
    H([I]),
    s,
    A,
    B,
    K
  ]);

  return Session(key: K.toHex(), proof: M.toHex());
}

verifySession(String clientPublicEphemeral, Session clientSession, String serverSessionProof) {
  final H = Params.H;
  final A = BigInt.parse(clientPublicEphemeral, radix: 16);
  final M = BigInt.parse(clientSession.proof, radix: 16);
  final K = BigInt.parse(clientSession.key, radix: 16);

  final expected = H([A, M, K]);
  final actual = BigInt.parse(serverSessionProof, radix: 16);

  if (actual != expected) {
    throw Exception('Server provided session proof is invalid');
  }
}
